game-core/src/main/java/games/strategy/engine/data/Resource.java

```diff
@@ -1,31 +1,25 @@
 package games.strategy.engine.data;
 
-import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 
 public class Resource extends NamedAttachable {
   private static final long serialVersionUID = 7471431759007499935L;
 
   private final List<PlayerID> players;
 
-  /**
-   * Creates new Resource.
-   *
-   * @param name name of the resource
-   * @param data game data
-   */
-  public Resource(final String name, final GameData data) {
-    this(name, data, new ArrayList<>());
+  public Resource(final String resourceName, final GameData data) {
+    this(resourceName, data, Collections.emptyList());
```

Comment:
Using `Collections#emptyList()` might cause surprises later because the three-parameter constructor doesn't make a defensive copy of this collection.  For example, if a future feature required adding a new element to this collection post-construction, it would fail because `Collections#emptyList()` returns an immutable collection.

I wouldn't recommend making a defensive copy of the `players` collection in the three-parameter constructor without a lot of testing.  Some legacy aspect of the code may depend on having a direct reference to the argument.

Actually, @ron-murhammer added this code recently in #3034.  He may have a better perspective on what is or is not safe to do here.

Reply:
> For example, if a future feature required adding a new element to this collection post-construction, it would fail because Collections#emptyList() returns an immutable collection.

Agree, but it's a hypothetical requirement at the moment. Our current requirements do not need that, so I would default to the simplest reasonable solution that fits our current needs. Then when we do get new requirements, we'll be in the best place to then work in those requirements, rather than undoing things we thought we needed.

In this case there should be no surprises as the variable usage is well encapsulated. We only need this variable for the `isDisplayedFor` check a few lines later, which is a read operation.

***********

Digging into the usage, there are only two callers to `new Resource(arg, arg, arg)`. One uses a locally scoped variable, another uses `data.getPlayerList().getPlayers()`, which creates a defensive copy already.

Reply:
I'm not sure I agree with this defensive copy, but it would make it clear that we are doing things which are 'safe' here, so I added a defensive copy to the three-arg constructor: 7c4e1c5 (it's not necessary, but I'm sure we'll probably wonder and check again about this, and the extra list references does not hurt)